src/transform/src/canonicalize_mfp.rs

```diff
@@ -61,55 +64,100 @@ impl CanonicalizeMfp {
-                // to
-                // `SELECT f1, f2, f3 FROM t, (SELECT * FROM (VALUES (lit1, lit2))) as filter_list
-                //  WHERE t.f1 = filter_list.column1 AND t.f2 = filter_list.column2`
-
-                let inp_id = id.clone();
-                let inp_typ = relation.typ();
-                let filter_list = MirRelationExpr::Constant {
-                    rows: Ok(vec![(val.clone(), 1)]),
-                    typ: mz_repr::RelationType {
-                        column_types: key.iter().map(|e| e.typ(&inp_typ.column_types)).collect(),
-                        keys: vec![],
-                    },
-                };
+            let orig_mfp = mfp.clone();
 
-                let join = MirRelationExpr::Join {
-                    inputs: vec![relation.clone().arrange_by(&[key.clone()]), filter_list],
-                    equivalences: key
-                        .iter()
-                        .enumerate()
-                        .map(|(i, e)| {
-                            vec![(*e).clone(), MirScalarExpr::column(i + inp_typ.arity())]
-                        })
-                        .collect(),
-                    implementation: IndexedFilter(inp_id, key.clone(), val),
-                };
+            // Prepare for the literal constraints detection. These might make the MFP more
+            // complicated, so we'll later want to either get back to orig_mfp, or undo these.
+            CanonicalizeMfp::inline_literal_constraints(&mut mfp)?;
+            CanonicalizeMfp::list_of_predicates_to_and_of_predicates(&mut mfp);
+            CanonicalizeMfp::distribute_and_over_or(&mut mfp)?;
 
-                *relation = MirRelationExpr::Project {
-                    input: Box::new(join),
-                    outputs: (0..inp_typ.arity()).collect(),
-                };
+            // todo: We might want to also call `canonicalize_equivalences`,
+            // see near the end of literal_constraints.slt.
+
+            let key_val = CanonicalizeMfp::detect_literal_constraints(&mfp, *id, indexes);
+
+            match key_val {
+                None => {
+                    // We didn't find a usable index, let's go with the original MFP.
+                    mfp = orig_mfp;
+                }
+                Some((key, possible_vals)) => {
+                    // We found a usable index. We'll try to remove the corresponding literal
+                    // constraint.
+                    if CanonicalizeMfp::remove_literal_constraints(&mut mfp, &key) {
+                        // We were able to remove the literal constraint, so we would like to use
+                        // this new MFP, so we try undoing the preparation for the literal
+                        // constraint detection.
+                        // undo distribute_and_over_or, list_of_predicates_to_and_of_predicates
+                        CanonicalizeMfp::canonicalize_predicates(&mut mfp, relation_cloned);
```

Comment:
It seems that `relation_cloned` is only used here, so can you pass a reference to `relation` to `canonicalize_predicates` and then have the helper method do the cloning?

Or at least move the cloning so that it only happens in this branch?

Reply:
Hmm, yes! It was possible after I changed the outer pattern match to not mut& borrow a part of `relation`.